File: benchmarks/pandas/bench_to_date_input.py
"""
Benchmark: pandas pd.Timestamp() — convert ISO strings, timestamps, and datetime objects to Timestamp.
Mirrors tsb bench_to_date_input.ts (toDateInput).
Outputs JSON: {"function": "to_date_input", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
from datetime import datetime
import pandas as pd

WARMUP = 5
ITERATIONS = 50

iso_strings = [
    "2020-01-01",
    "2024-03-15T10:30:00Z",
    "2023-12-31T23:59:59.999Z",
    "2022-07-04",
    "2021-01-01T00:00:00",
]

# millisecond timestamps
timestamps = [
    0,
    1_577_836_800_000,
    1_704_067_200_000,
    1_609_459_200_000,
    1_672_531_200_000,
]

date_objects = [
    datetime(2020, 1, 1),
    datetime(2024, 6, 15),
    datetime(2001, 9, 9, 1, 46, 40),
]

SIZE = 10_000
str_batch = [
    f"{2000 + (i % 25)}-{((i % 12) + 1):02d}-{((i % 28) + 1):02d}"
    for i in range(SIZE)
]
num_batch = [i * 86_400_000 for i in range(SIZE)]

for _ in range(WARMUP):
    for s in iso_strings:
        pd.Timestamp(s)
    for t in timestamps:
        pd.Timestamp(t, unit="ms")
    for d in date_objects:
        pd.Timestamp(d)

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    for _ in range(1000):
        for s in iso_strings:
            pd.Timestamp(s)
        for t in timestamps:
            pd.Timestamp(t, unit="ms")
        for d in date_objects:
            pd.Timestamp(d)
    for s in str_batch:
        pd.Timestamp(s)
    for t in num_batch:
        pd.Timestamp(t, unit="ms")
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
print(json.dumps({
    "function": "to_date_input",
    "mean_ms": round(total_ms / ITERATIONS, 3),
    "iterations": ITERATIONS,
    "total_ms": round(total_ms, 3),
}))

File: benchmarks/tsb/bench_to_date_input.ts
/**
 * Benchmark: toDateInput — convert ISO strings, timestamps, and Date objects to Date.
 * Mirrors pandas pd.Timestamp() single-value date parsing.
 * Outputs JSON: {"function": "to_date_input", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { toDateInput } from "../../src/index.ts";

const WARMUP = 5;
const ITERATIONS = 50;

const isoStrings = [
  "2020-01-01",
  "2024-03-15T10:30:00Z",
  "2023-12-31T23:59:59.999Z",
  "2022-07-04",
  "2021-01-01T00:00:00",
];

const timestamps = [
  0,
  1_577_836_800_000, // 2020-01-01
  1_704_067_200_000, // 2024-01-01
  1_609_459_200_000, // 2021-01-01
  1_672_531_200_000, // 2023-01-01
];

const dateObjects = [
  new Date("2020-01-01"),
  new Date("2024-06-15"),
  new Date(1_000_000_000_000),
];

const SIZE = 10_000;
const strBatch = Array.from({ length: SIZE }, (_, i) => {
  const y = 2000 + (i % 25);
  const m = (i % 12) + 1;
  const d = (i % 28) + 1;
  return `${y}-${m.toString().padStart(2, "0")}-${d.toString().padStart(2, "0")}`;
});
const numBatch = Array.from({ length: SIZE }, (_, i) => i * 86_400_000);

for (let w = 0; w < WARMUP; w++) {
  for (const s of isoStrings) toDateInput(s);
  for (const t of timestamps) toDateInput(t);
  for (const d of dateObjects) toDateInput(d);
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const t0 = performance.now();
  for (let j = 0; j < 1000; j++) {
    for (const s of isoStrings) toDateInput(s);
    for (const t of timestamps) toDateInput(t);
    for (const d of dateObjects) toDateInput(d);
  }
  for (const s of strBatch) toDateInput(s);
  for (const t of numBatch) toDateInput(t);
  times.push(performance.now() - t0);
}

const total = times.reduce((a, b) => a + b, 0);
console.log(
  JSON.stringify({
    function: "to_date_input",
    mean_ms: round3(total / ITERATIONS),
    iterations: ITERATIONS,
    total_ms: round3(total),
  }),
);

function round3(v: number): number {
  return Math.round(v * 1000) / 1000;
}
